packages/clerk-js/src/utils/queryStateParams.ts

```diff
@@ -0,0 +1,40 @@
+import { getClerkQueryParam } from '../utils';
+
+export const decodeBase64Json = (value: string) => {
+  return JSON.parse(atob(value));
+};
+
+export const encodeBase64Json = (value: { path: string; componentName: string }) => {
+  return btoa(JSON.stringify(value));
+};
+
+export const readAndRemoveStateParam = () => {
+  const urlClerkState = getClerkQueryParam('__clerk_state') ?? '';
+
+  return urlClerkState ? decodeBase64Json(urlClerkState) : null;
+};
+
+type SerializeAndAppendModalStateProps = { url: string; currentPath: string; componentName: string };
+export const serializeAndAppendModalState = ({
+  url,
+  currentPath,
+  componentName,
+}: SerializeAndAppendModalStateProps) => {
+  const regexPattern = /CLERK-ROUTER\/VIRTUAL\/.*\//;
+
+  const redirectParams = {
+    path: currentPath.replace(regexPattern, '') || '',
+    componentName,
+  };
+
+  const encodedRedirectParams = encodeBase64Json(redirectParams);
+
+  const urlWithParams = new URL(url);
+  const searchParams = urlWithParams.searchParams;
+
+  searchParams.set('__clerk_state', encodedRedirectParams);
```

Comment:
⛔ Let's rename the param to `__clerk_modal_state` as we already have a `__clerk_status` param for authentication purposes and the name will cause confusion to the future reader.